File: frontend/src/services/navigation.ts
import api from './api';

export interface RoleBrief {
  id: number;
  name: string;
}

export interface NavigationLink {
  id: number;
  category: string;
  name: string;
  url: string;
  icon?: string;
  description?: string;
  sortOrder: number;
  isActive: boolean;
  roles: RoleBrief[];
  createdAt: string;
  updatedAt: string;
}

export interface NavigationGroup {
  category: string;
  links: NavigationLink[];
}

export interface NavigationLinkListResponse {
  items: NavigationLink[];
  total: number;
  page: number;
  page_size: number;
}

export interface NavigationLinkCreate {
  category: string;
  name: string;
  url: string;
  icon?: string;
  description?: string;
  sortOrder?: number;
  isActive?: boolean;
  restrictToCurrentRole?: boolean;
}

export interface NavigationLinkUpdate {
  category?: string;
  name?: string;
  url?: string;
  icon?: string;
  description?: string;
  sortOrder?: number;
  isActive?: boolean;
  restrictToCurrentRole?: boolean;
}

export interface NavigationLinkImportResult {
  success: boolean;
  name: string;
  message: string;
}

export interface NavigationImportResponse {
  total: number;
  success_count: number;
  failed_count: number;
  results: NavigationLinkImportResult[];
}

export interface NavigationLinkImport {
  category: string;
  name: string;
  url: string;
  icon?: string;
  description?: string;
  sortOrder?: number;
  isActive?: boolean;
  roleNames?: string;
}

export const navigationApi = {
  getPublicLinks: async (): Promise<{ groups: NavigationGroup[] }> => {
    const response = await api.get('/navigation/public');
    return response.data;
  },

  getLinks: async (params?: {
    category?: string;
    is_active?: boolean;
    page?: number;
    page_size?: number;
  }): Promise<NavigationLinkListResponse> => {
    const response = await api.get('/navigation', { params });
    return response.data;
  },

  getLink: async (id: number): Promise<NavigationLink> => {
    const response = await api.get(`/navigation/${id}`);
    return response.data;
  },

  createLink: async (data: NavigationLinkCreate): Promise<NavigationLink> => {
    const response = await api.post('/navigation', data);
    return response.data;
  },

  updateLink: async (id: number, data: NavigationLinkUpdate): Promise<NavigationLink> => {
    const response = await api.put(`/navigation/${id}`, data);
    return response.data;
  },

  deleteLink: async (id: number): Promise<void> => {
    await api.delete(`/navigation/${id}`);
  },

  exportLinks: async (): Promise<void> => {
    const response = await api.get('/navigation/export', {
      responseType: 'blob',
    });
    const blob = new Blob([response.data], { type: 'text/csv;charset=utf-8' });
    const url = window.URL.createObjectURL(blob);
    const link = document.createElement('a');
    link.href = url;
    const contentDisposition = response.headers['content-disposition'];
    let filename = 'navigation_export.csv';
    if (contentDisposition) {
      const filenameMatch = contentDisposition.match(/filename=(.+)/);
      if (filenameMatch) {
        filename = filenameMatch[1];
      }
    }
    link.download = filename;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    window.URL.revokeObjectURL(url);
  },

  importLinks: async (data: NavigationLinkImport[]): Promise<NavigationImportResponse> => {
    const response = await api.post('/navigation/import', data);
    return response.data;
  },
};

File: backend/app/schemas/navigation.py
"""
Navigation link schemas.
"""

from app.schemas.base import BaseResponse, BaseSchema


class NavigationLinkCreate(BaseSchema):
    """Schema for creating a navigation link."""

    category: str
    name: str
    url: str
    icon: str | None = None
    description: str | None = None
    sort_order: int = 0
    is_active: bool = True
    restrict_to_current_role: bool = True


class NavigationLinkUpdate(BaseSchema):
    """Schema for updating a navigation link."""

    category: str | None = None
    name: str | None = None
    url: str | None = None
    icon: str | None = None
    description: str | None = None
    sort_order: int | None = None
    is_active: bool | None = None
    restrict_to_current_role: bool | None = None


class RoleBrief(BaseSchema):
    """Brief role info for navigation link response."""

    id: int
    name: str


class NavigationLinkResponse(BaseResponse):
    """Schema for navigation link response."""

    category: str
    name: str
    url: str
    icon: str | None = None
    description: str | None = None
    sort_order: int
    is_active: bool
    roles: list[RoleBrief] = []


class NavigationLinkListResponse(BaseSchema):
    """Schema for navigation link list response."""

    items: list[NavigationLinkResponse]
    total: int


class NavigationLinkGrouped(BaseSchema):
    """Schema for navigation links grouped by category."""

    category: str
    links: list[NavigationLinkResponse]


class NavigationLinkImport(BaseSchema):
    """Schema for importing a navigation link from CSV."""

    category: str
    name: str
    url: str
    icon: str | None = None
    description: str | None = None
    sort_order: int = 0
    is_active: bool = True
    role_names: str | None = None


class NavigationLinkImportResult(BaseSchema):
    """Schema for import result."""

    success: bool
    name: str
    message: str


class NavigationImportResponse(BaseSchema):
    """Schema for bulk import response."""

    total: int
    success_count: int
    failed_count: int
    results: list[NavigationLinkImportResult]
